lua/lspconfig/denols.lua

```diff
@@ -1,8 +1,76 @@
 local configs = require 'lspconfig/configs'
 local util = require 'lspconfig/util'
+local lsp = vim.lsp
 
 local server_name = "denols"
 
+local function buf_cache(bufnr)
+  local params = {}
+  params['referrer'] = { textDocument = { uri = vim.uri_from_bufnr(bufnr) } }
+  params['uris'] = {}
+  lsp.buf_request(0, 'deno/cache', params)
+end
+
+local function split(str, delimiter)
+  local result = { }
+  local from  = 1
+  local delim_from, delim_to = string.find( str, delimiter, from  )
+  while delim_from do
+    table.insert( result, string.sub( str, from , delim_from-1 ) )
+    from  = delim_to + string.len(delimiter)
+    delim_from, delim_to = string.find( str, delimiter, from  )
+  end
+  table.insert( result, string.sub( str, from  ) )
+  return result
+end
+
+local function virtual_text_document_handler(uri, result)
+  if not result or #result ~= 1 then return nil end
+
+  local lines = split(result[1].result, '\n')
+  local bufnr = vim.uri_to_bufnr(uri)
+
+  local current_buf = vim.api.nvim_buf_get_lines(bufnr, 0, -1, false)
+  if #current_buf ~= 0 then return nil end
+
+  vim.api.nvim_buf_set_lines(bufnr, 0, -1, nil, lines)
+  vim.api.nvim_buf_call(bufnr, function()
+    vim.cmd('set readonly')
+    vim.cmd('set nomodified')
+  end)
+end
+
+local function virtual_text_document(uri)
+  local params = {
+    textDocument = {
+      uri = uri,
+    },
+  }
+  local result = lsp.buf_request_sync(0, 'deno/virtualTextDocument', params)
+  virtual_text_document_handler(uri, result)
+end
+
+local default_definition_handler = lsp.handlers['textDocument/definition']
+local function deno_definition_handler(_, method, result)
+  if not result or vim.tbl_isempty(result) then return nil end
+
+  for _, res in pairs(result) do
+    if string.sub(res.targetUri, 1, 7) == 'deno://' then
+      virtual_text_document(res.targetUri)
+    end
+  end
+
+  default_definition_handler(_, method, result)
+end
+
+local function deno_definition()
+  lsp.handlers['textDocument/definition'] = function(m, method, result)
+    deno_definition_handler(m, method, result)
+    lsp.handlers['textDocument/definition'] = default_definition_handler
+  end
+  lsp.buf.definition()
+end
```

Comment:
Seems a bit fragile/dangerous/dirty to juggle the handlers like this 🤔

Reply:
Hmm. Would you like the followings?

```lua
configs[server_name] = {
  default_config = {
    ...
    handlers = {
      ['textDocument/definition'] = function(err, method, result)
        if not result or vim.tbl_isempty(result) then return nil end

        for _, res in pairs(result) do
          if string.sub(res.targetUri, 1, 7) == 'deno://' then
            virtual_text_document(res.targetUri)
          end
        end

        vim.lsp.handlers['textDocument/definition'](err, method, result)
      end;
    };
  };
```

Reply:
Well, it's not up to me. It was mostly an observation. But _I_ definitely prefer overriding server-specific handler rather than modifying the global one (like your example above).

Another alternative is to leave the `textDocument/definition` handler alone and implement everything in the `deno_definition` function. Something like this:

```lua
local util = require('vim.lsp.util')

local function deno_definition()
  local params = util.make_position_params()

  vim.lsp.buf_request(0, 'textDocument/definition', params, function(...)
    local result = select(3, ...) or {}

    for _, res in pairs(result) do
      if string.sub(res.targetUri, 1, 7) == 'deno://' then
        virtual_text_document(res.targetUri)
      end
    end

    vim.lsp.handlers['textDocument/definition'](...)
  end)
end
```